### PR TITLE
IOS-7746 Sheet v2

### PR DESCRIPTION
**Pull request by jmbrocal**
Hi folks!!

This PR contains the evolution of the Sheet component.

- Update the Sheet v1 implementation to support two list types: splitting the code in SheetViewController.
- A few fixes, (thank you @jmpg93 for helping me!!!)
- Implement two new list types: Actions and Informative
- Update Sheet catalog so we can test the new list types.
- Regenerate reference images of Feedback component.

Tests of Sheet were disabled because they are currently broken in main, I will fix it in another PR.

Documentation: https://confluence.tid.es/x/qNjLC
Figma: https://www.figma.com/file/nxToAYk7FGC86TS0mY61TU/Sheet?node-id=804%3A4932

You can test the component using the following build of the catalog:

image

### CODE REVIEW

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 1]
```diff
@@ -78,6 +81,38 @@ class UICatalogSheetViewController: UIViewController {
...
+        cell.segmentedControl.insertSegment(withTitle: "small", at: 2, animated: false)
+
+        cell.segmentedControl.selectedSegmentIndex = 0
+        return cell
+    }()
+
+    private lazy var actionStyleCell: UISegmentedControlTableViewCell = {
+        let cell = UISegmentedControlTableViewCell(reuseIdentifier: "actionStyle")
+
+        cell.segmentedControl.insertSegment(withTitle: "default", at: 0, animated: false)
```
Comment (alejandroruizponce) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:109]:
```suggestion
        cell.segmentedControl.insertSegment(withTitle: "Default", at: 0, animated: false)
```

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 2]
```diff
@@ -78,6 +81,38 @@ class UICatalogSheetViewController: UIViewController {
...
+        cell.segmentedControl.insertSegment(withTitle: "destructive", at: 1, animated: false)
```
Comment (alejandroruizponce) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:110]:
```suggestion
        cell.segmentedControl.insertSegment(withTitle: "Destructive", at: 1, animated: false)
```

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 3]
```diff
@@ -78,6 +81,38 @@ class UICatalogSheetViewController: UIViewController {
...
+        cell.segmentedControl.insertSegment(withTitle: "Action", at: 2, animated: false)
+
+        cell.segmentedControl.selectedSegmentIndex = 0
+        return cell
+    }()
+
+    private lazy var informativeIconTypeCell: UISegmentedControlTableViewCell = {
+        let cell = UISegmentedControlTableViewCell(reuseIdentifier: "informativeIconType")
+
+        cell.segmentedControl.insertSegment(withTitle: "bullet", at: 0, animated: false)
```
Comment (alejandroruizponce) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:98]:
```suggestion
        cell.segmentedControl.insertSegment(withTitle: "Bullet", at: 0, animated: false)
```

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 4]
```diff
@@ -187,37 +271,90 @@ extension UICatalogSheetViewController: UITableViewDataSource, UITableViewDelega
...
-            print("Sheet selection response:")
-            print("\tAction:\(sheetResponse.action)")
-            print("\tSelected ids:\(sheetResponse.selectedIds)")
+        return configuration
+    }
+
+    var informativeSheet: SheetConfiguration {
+        var rows: [InformativeItem] = []
+
+        let icon: InformativeItemIcon
```
Comment (alejandroruizponce) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:280]: Are you not checking at any time whether the icon should be visible or not? It's always visible in Catalog for informative or action lists.

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 5]
```diff
@@ -187,37 +271,90 @@ extension UICatalogSheetViewController: UITableViewDataSource, UITableViewDelega
...
+        var rows: [ActionItem] = []
 
-    var sheetNumElements: Int {
-        Int(numberOfElementsCell.currentValue)
+        for index in 1 ... sheetNumElements {
+            let item = ActionItem(
+                id: index.description,
+                title: "List Item \(index)",
+                style: actionStyleCell.segmentedControl.selectedSegmentIndex == 0 ? .normal : .destructive,
+                url: "https://img.icons8.com/ios-glyphs/344/bookmark.png",
```
Comment (alejandroruizponce) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:335]: Same here, icon has to be always visible in this list type? Here appears without icon:
https://www.figma.com/file/nxToAYk7FGC86TS0mY61TU/Sheet?node-id=1960%3A13985

File: Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift [hunk 6]
```diff
@@ -63,7 +63,7 @@ class SheetPresentationController: UIPresentationController {
         }
 
         /// The maximum height allowed for the sheet. We allow the sheet to reach the top safe area inset.
-        let maximumHeight = containerView.frame.height - containerView.safeAreaInsets.top - containerView.safeAreaInsets.bottom
```
Comment (alejandroruizponce) [inline Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift:66]: Cool!

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift [hunk 7]
```diff
@@ -367,18 +249,36 @@ private extension SheetViewController {
...
+                guard let s = self else { return }
+                s.dismiss(animated: true)
             }
         }
     }
 }
+
+private class FittingCompressedScrollStackView: ScrollStackView {
+    override var intrinsicContentSize: CGSize {
+        stackView.systemLayoutSizeFitting(
```
Comment (alejandroruizponce) [inline Sources/Mistica/Components/Sheet/View/SheetViewController.swift:275]: Good approach! But I find some issues in these cases:

11 action rows in sheet with title:


11 action rows without title:

**Review by alejandroruizponce — CHANGES_REQUESTED** [review]

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift [hunk 8]
```diff
@@ -367,18 +249,36 @@ private extension SheetViewController {
...
-                }
+    func handleListRowTapped(_ sheetList: SheetList, rowTapped: ListFragmentView.ItemTappedType) {
+        if !isDismissing && sheetList.autoSubmit {
+            isDismissing = true
+
+            switch rowTapped {
+            case .action(let item):
+                sheetSelectionResponse = .init(action: .submit, selectedIds: [.init(id: sheetList.id, selected: [item.id])])
+            case .informative:
+                sheetSelectionResponse = .init(action: .submit, selectedIds: [])
```
Comment (alejandroruizponce) [inline Sources/Mistica/Components/Sheet/View/SheetViewController.swift:260]: Mm, maybe it comes in the documentation but I have not seen it, when taping in a row informative type, the idea is that we do not return that id row selected or anything? Only we dismiss the sheet? And if so, when testing I have seen that the action that is sent is SUBMIT instead of DISMISS. Is this correct?

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift (hunk 3 above)
Comment (jmbrocal) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:98]: Fixed this and other segments

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift (hunk 2 above)
Comment (jmbrocal) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:110]: done here and other segment

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift (hunk 5 above)
Comment (jmbrocal) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:335]: Nop, we support actions without images. I will refactor this screen in other PR. Ty

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift (hunk 8 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/SheetViewController.swift:260]: About the selection: yes. For informative type we do not return the row tapped. About the submit or dismiss, I am not sure but I will check it with android.

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift (hunk 7 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/SheetViewController.swift:275]: The whole structure of this component is really hard to follow... I will investigate this case...

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift (hunk 4 above)
Comment (jmbrocal) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:280]: Informative always has icons while actions can be displayed without icons.

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+//  ActionRow.swift
+//
+//  Made with ❤️ by Novum
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import UIKit
+
+class ActionRow: UIStackView {
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift:11]: Minor, but this is the only view that inherits from `UIStackView` even tho every other one is also using `UIStackView` as the primary content view.

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        ])
+        addSubview(withDefaultConstraints: frontStackView)
+
+        backStackView.addArrangedSubview(iconImageView)
+        backStackView.addArrangedSubview(dummyTitleLabel)
+
+        frontStackView.addArrangedSubview(dummyView)
+        frontStackView.addArrangedSubview(titleLabel)
+
+        dummyView.backgroundColor = .red
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift:179]: We don't need this.
```suggestion
```

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/SingleSelectionRow.swift [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            radioButton.isActivated
+        }
+        set {
+            radioButton.isActivated = newValue
+        }
+    }
+}
+
+private extension SingleSelectionRowView {
+    func setUp() {
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/SingleSelectionRow.swift:78]:
```suggestion
        commonInit()
    }

    @available(*, unavailable)
    required init(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension SingleSelectionRowView {
    var isSelected: Bool {
        get {
            radioButton.isActivated
        }
        set {
            radioButton.isActivated = newValue
        }
    }
}

private extension SingleSelectionRowView {
    func commonInit() {
```

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift [hunk 12]
```diff
@@ -208,121 +197,24 @@ public extension SheetViewController {
...
         }
 
-        // Set sheet maximum height in 70% of device screen height.
-        let scrollHeightMax = containerView.heightAnchor.constraint(lessThanOrEqualToConstant: UIScreen.main.bounds.height * 0.7)
-        scrollHeightMax.priority = .required
-        scrollHeightMax.isActive = true
-
-        let scrollHeight = scrollView.heightAnchor.constraint(equalTo: contentStackView.heightAnchor)
-        scrollHeight.priority = .defaultHigh
-        scrollHeight.isActive = true
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/SheetViewController.swift:305]: The height of the sheet is now fully manageb by the presentation controller.

File: Sources/Mistica/Utils/Views/ScrollStackView.swift [hunk 13]
```diff
@@ -33,6 +33,17 @@ class ScrollStackView: UIScrollView {
...
+    override var intrinsicContentSize: CGSize {
+        stackView.systemLayoutSizeFitting(
+            CGSize(
+                width: UIScreen.main.bounds.width,
+                height: UIView.layoutFittingCompressedSize.height
+            ),
+            withHorizontalFittingPriority: .required,
+            verticalFittingPriority: .defaultLow
+        )
+    }
```
Comment (jmpg93) [inline Sources/Mistica/Utils/Views/ScrollStackView.swift:45]: I think we can remove this. I created the `FittingCompressedScrollStackView` to avoid impacting other views.

File: Tests/MisticaTests/UI/FeedbackTests.swift [hunk 14]
```diff
@@ -34,7 +34,7 @@ final class FeedbackTests: XCTestCase {
 
     func testSuccessTitle() {
         assertSnapshotForAllBrandsAndStyles(
-            as: .image(on: .iPhoneSe),
+            as: .image,
```
Comment (jmpg93) [inline Tests/MisticaTests/UI/FeedbackTests.swift:37]: Why did you change this?

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift (hunk 8 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/SheetViewController.swift:260]: I updated the PR, now for the informative type the library returns action == dismiss. Ty!

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift (hunk 9 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift:11]: Yep, it was intentionally. The Action row is the most simple one so I just inherit from UIStackView so the view stack is more flattern

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    func load(icon: InformativeItemIcon, in imageView: UIImageView) {
+        switch icon {
+        case .bullet:
+            imageView.image = UIImage.circle(diameter: 6, color: .neutralHigh)
```
Comment (aweell) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift:183]: It seems that the bullet diameter is following the spec correctly, but when comparing the rendered component with bullet and an small icon (16px) they seem to have almost the same height:

Screenshot 2022-10-20 at 15 36 49

Of course the icon can be 6px height inside a 16px... I was checking the icon bookmark used in the catalog preview and the height of the icon inside the 16px box is ≈14px. 

I tried proportionally scaling the Figma definition to the screenshot size and the bullet still seems smaller: 

Screenshot 2022-10-20 at 15 51 28

In addition, even though the neutralHigh is specified as the bullet color, it's not working as expected in dark mode:
 
Screenshot 2022-10-20 at 16 16 23

```
let neutralHigh = MovistarColors.palette.grey6 | MovistarColors.palette.grey2
```

The `palette.grey2` color should be `#eeeeee` but it doesn't seems to be the color displayed...

File: Sources/Mistica/Components/Sheet/View/Fragments/List/ListFragmentView.swift [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    func fillListWithActionItems(_ items: [ActionItem]) {
+        for item in items {
+            let rowView = ActionRow(item: item)
+
+            let itemTapGesture = UITapGestureRecognizer(
+                target: self,
+                action: #selector(selectItem(_:))
+            )
+            rowView.addGestureRecognizer(itemTapGesture)
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/ListFragmentView.swift:106]: Sad we are using a tap gesture. This works like a button so it would been nice to have some feedback on press.

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift (hunk 7 above)
Comment (jmpg93) [reply Sources/Mistica/Components/Sheet/View/SheetViewController.swift:276]: I can't reproduce this bug. How did you exactly do it??

**Review by alejandroruizponce — APPROVED** [review]
Great work!

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        if let url = item.url {
+            load(url: url, urlDark: item.urlDark, in: iconImageView)
+            iconImageView.tintColor = imageTintColor
+        } else {
+            iconImageView.isHidden = true
+        }
+    }
+
+    func layoutViews() {
+        spacing = 16
```
Comment (amegias) [inline Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift:70]: is there any spacing between actions??
Captura de Pantalla 2022-10-20 a las 16 26 02

Captura de Pantalla 2022-10-20 a las 17 58 13

**Review by amegias — APPROVED** [review]

**Comment by jmbrocal** [comment]
Hi folks!

I have to update the PR with
- Adding support for changing the background row while it is pressed (it is now defined in the spec)
- Fixing some ui issues

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift (hunk 17 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/ActionRow.swift:70]: I noticed this bug, the PR was updated

File: Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift (hunk 15 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift:185]: Hey!

About the bullet icon: I am not using an asset, I am drawing it from code. I was using a diameter of 6 (i think it is the size in figma) but I will fix to use a diameter of 14. it is ok or it should be 16?

about the color in dark move, I will review it

Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/Fragments/List/Rows/InformativeRow.swift]: Talked about his in private. Fixed both cases. Ty!

**Comment by jmbrocal** [comment]
More fixes, the last one I hope...

Add bullet icon

File: Sources/Mistica/Components/Sheet/View/SheetViewController.swift (hunk 7 above)
Comment (jmbrocal) [reply Sources/Mistica/Components/Sheet/View/SheetViewController.swift:278]: It is already fixed (but I have not idea how I fixed it XD)

File: Sources/Mistica/Utils/UIKitPreviewProvider/ViewPreview.swift [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+//  Made with ❤️ by Novum
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import SwiftUI
+import UIKit
+
+@available(iOS 13.0.0, *)
+struct ViewPreview: UIViewRepresentable {
```
[the remark this answers is not in the document]
Comment (jmbrocal) [reply Sources/Mistica/Utils/UIKitPreviewProvider/ViewPreview.swift:13]: Ups, not needed. I used it while implementing the new rows (so I can use the SwiftUI Previews) but we cannot use this bc we must support iOS 13. Removing!

**Review by jmpg93 — APPROVED** [review]

**Review by amegias — APPROVED** [review]

**Review by alejandroruizponce — APPROVED** [review]

**Comment by aweell** [comment]
It seems that the informative `list-item` label and the actions label are not using the correct text-preset, both should use `.textPreset3(weight: .regular)` causing those texts to render even bigger than the sheet title.

It can be compared with the single selection that is using the correct `font-size`.

From the specs:

* Informative list text-presets
Screenshot 2022-10-21 at 11 53 38

* Actions list text-presets
Screenshot 2022-10-21 at 11 53 46

**Comment by jmbrocal** [comment]
> It seems that the informative `list-item` label and the actions label are not using the correct text-preset, both should use `.textPreset3(weight: .regular)` causing those texts to render even bigger than the sheet title.
> 
> It can be compared with the single selection that is using the correct `font-size`.
> 
> From the specs:
> 
> * Informative list text-presets
>   Screenshot 2022-10-21 at 11 53 38
> * Actions list text-presets
>   Screenshot 2022-10-21 at 11 53 46

good catch, the font was missing for that row. fixed
